File: Water_and_Jug_Problem.cpp
class Solution {
public:
    bool canMeasureWater(int a, int b, int target) {
        set<pair<int,int>> seen;
        queue<pair<int,int>> q;
        q.push({0,0});
        seen.insert({0,0});
        while(!q.empty()){
            pair<int,int> curr = q.front();
            q.pop();
            pair<int,int> temp;
            temp = fillA(curr,a,b);
            if(temp.first==target || temp.second==target) return true;
            if(!seen.contains(temp)) q.push(temp);
            temp = fillB(curr,a,b);
            if(temp.first==target || temp.second==target) return true;
            if(!seen.contains(temp)) q.push(temp);
            temp = emptyA(curr,a,b);
            if(temp.first==target || temp.second==target) return true;
            if(!seen.contains(temp)) q.push(temp);
            temp = emptyB(curr,a,b);
            if(temp.first==target || temp.second==target) return true;
            if(!seen.contains(temp)) q.push(temp);
            temp = aTob(curr,a,b);
            if(temp.first==target || temp.second==target) return true;
            if(!seen.contains(temp)) q.push(temp);
            temp = bToa(curr,a,b);
            if(temp.first==target || temp.second==target) return true;
            if(!seen.contains(temp)) q.push(temp);
        }
        return false;
    }
    pair<int,int> fillA(pair<int,int> curr, int a, int b){
        int x = curr.first, y = curr.second;
        return {a,y};
    }
    pair<int,int> fillB(pair<int,int> curr, int a, int b){
        int x = curr.first, y = curr.second;
        return {x,b};
    }
    pair<int,int> emptyA(pair<int,int> curr, int a, int b){
        int x = curr.first, y = curr.second;
        return {0,y};
    }
    pair<int,int> emptyB(pair<int,int> curr, int a, int b){
        int x = curr.first, y = curr.second;
        return {x,0};
    }
    pair<int,int> aTob(pair<int,int> curr, int a, int b){
        int x = curr.first, y = curr.second;
        if(b-y >= x) return {0,x+y};
        else return {x-b+y , b};
    }
    pair<int,int> bToa(pair<int,int> curr, int a, int b){
        int x = curr.first, y = curr.second;
        if(a-x >= y) return {x+y,0};
        else return {a, y-a+x};
    }
};

File: Water_and_Jug_Problem.py
class Solution:
    def canMeasureWater(self, a: int, b: int, target: int) -> bool:
        seen = set()
        q = deque([(0, 0)])
        seen.add((0, 0))
        
        while q:
            x, y = q.popleft()
            
            if x + y == target:
                return True
            
            next_options = [
                self.fillA(x, y, a, b),
                self.fillB(x, y, a, b),
                self.emptyA(x, y, a, b),
                self.emptyB(x, y, a, b),
                self.aTob(x, y, a, b),
                self.bToa(x, y, a, b)
            ]
            
            for temp in next_options:
                if temp not in seen:
                    q.append(temp)
                    seen.add(temp)
                    
        return False

    def fillA(self, x: int, y: int, a: int, b: int) -> tuple:
        return (a, y)

    def fillB(self, x: int, y: int, a: int, b: int) -> tuple:
        return (x, b)

    def emptyA(self, x: int, y: int, a: int, b: int) -> tuple:
        return (0, y)

    def emptyB(self, x: int, y: int, a: int, b: int) -> tuple:
        return (x, 0)

    def aTob(self, x: int, y: int, a: int, b: int) -> tuple:
        if b - y >= x:
            return (0, x + y)
        else:
            return (x - (b - y), b)

    def bToa(self, x: int, y: int, a: int, b: int) -> tuple:
        if a - x >= y:
            return (x + y, 0)
        else:
            return (a, y - (a - x))
// Title: Water and Jug Problem
